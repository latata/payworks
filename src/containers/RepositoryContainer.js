import { Container } from 'unstated';

import repositoriesService from '../service/repositories';

export default class RepositoryContainer extends Container {
  state = {
    branches: [],
    status: 'uninitialized',
  };

  async fetchData(fullName) {
    this.setState({
      branches: [],
      status: 'loading',
    });

    try {
      const branches = await repositoriesService.fetchBranches(fullName);

      this.setState({
        branches,
        status: 'loaded',
      })
    } catch (error) {
      if (error.response && error.response.status === 404) {
        this.setState({
          status: 'not-found',
        });
      } else {
        this.setState({
          status: 'error',
        })
      }
    }
  };
}
