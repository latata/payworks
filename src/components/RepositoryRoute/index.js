import React, { useEffect } from 'react';
import connect from 'unstated-connect';
import RepositoryContainer from '../../containers/RepositoryContainer';
import Spinner from '../Spinner';

function RepositoryRoute({ match, containers: [repositoryContainer] }) {
  const fullName = match.params.id;
  const { state: { branches, status } } = repositoryContainer;

  useEffect(() => {
    repositoryContainer.fetchData(fullName);
  }, [fullName, repositoryContainer]);

  return (
    <div className="container">
      <h1>
        {fullName}
        {' '}
        <small className="text-muted">branches</small>
      </h1>
      {status === 'loaded' && (
        <ul>{branches.map((branch) => <li key={branch.name}><a
          href={`https://github.com/${fullName}/tree/${branch.name}`}>{branch.name}</a></li>)}</ul>
      )}
      {status === 'loading' && <Spinner />}
      {status === 'not-found' && <div>Repository does not exist</div>}
      {status === 'error' && <div>Unexpected error</div>}
    </div>
  );
}

export default connect([RepositoryContainer])(RepositoryRoute);
