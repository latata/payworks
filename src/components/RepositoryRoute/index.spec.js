import { cleanup, render } from 'react-testing-library';
import { Provider } from 'unstated';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import axios from 'axios';
import AxiosMockAdapter from 'axios-mock-adapter';

import RepositoryRoute from '../../components/RepositoryRoute';
import RepositoryContainer from '../../containers/RepositoryContainer';

const axiosMock = new AxiosMockAdapter(axios);

const flushPromises = () => new Promise(resolve => setTimeout(resolve));

function renderComponent(repositoryContainer) {
  return render(
    <Provider inject={[repositoryContainer]}>
      <MemoryRouter>
        <RepositoryRoute match={{
          params: {
            fullName: 'test-org/repo'
          }
        }} />
      </MemoryRouter>
    </Provider>
  );
}

afterEach(() => {
  cleanup();
  axiosMock.reset();
});

test('RepositoryRoute fetches branches on mount', async () => {
  axiosMock.onAny().reply(200, [
    { name: 'test-branch-1' },
    { name: 'test-branch-2' },
    { name: 'test-branch-3' },
  ]);

  const repositoryContainer = new RepositoryContainer();
  const { asFragment } = renderComponent(repositoryContainer);

  expect(asFragment()).toMatchSnapshot();

  await flushPromises();

  expect(asFragment()).toMatchSnapshot();
});

test('RepositoryRoute shows error message for repository which does not exist', async () => {
  axiosMock.onAny().reply(404);

  const repositoryContainer = new RepositoryContainer();
  const { asFragment } = renderComponent(repositoryContainer);

  await flushPromises();

  expect(asFragment()).toMatchSnapshot();
});

test('RepositoryRoute shows error message for 500 API response', async () => {
  axiosMock.onAny().reply(500);

  const repositoryContainer = new RepositoryContainer();
  const { asFragment } = renderComponent(repositoryContainer);

  await flushPromises();

  expect(asFragment()).toMatchSnapshot();
});
