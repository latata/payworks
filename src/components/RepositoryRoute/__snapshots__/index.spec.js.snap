// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryRoute fetches branches on mount 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      test-org/repo 
      <small
        class="text-muted"
      >
        branches
      </small>
    </h1>
  </div>
</DocumentFragment>
`;

exports[`RepositoryRoute fetches branches on mount 2`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      test-org/repo 
      <small
        class="text-muted"
      >
        branches
      </small>
    </h1>
    <ul>
      <li>
        <a
          href="https://github.com/test-org/repo/tree/test-branch-1"
        >
          test-branch-1
        </a>
      </li>
      <li>
        <a
          href="https://github.com/test-org/repo/tree/test-branch-2"
        >
          test-branch-2
        </a>
      </li>
      <li>
        <a
          href="https://github.com/test-org/repo/tree/test-branch-3"
        >
          test-branch-3
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`RepositoryRoute shows error message for 500 API response 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      test-org/repo 
      <small
        class="text-muted"
      >
        branches
      </small>
    </h1>
    <div>
      Unexpected error
    </div>
  </div>
</DocumentFragment>
`;

exports[`RepositoryRoute shows error message for repository which does not exist 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h1>
      test-org/repo 
      <small
        class="text-muted"
      >
        branches
      </small>
    </h1>
    <div>
      Repository does not exist
    </div>
  </div>
</DocumentFragment>
`;
